File: examples/csharp/avalonia/NotesShowcase.Tests/ViewModels/WorkspaceVMTests.cs
using NotesShowcase.Models;
using NotesShowcase.ViewModels;
using VMx.Dialogs;
using VMx.Lifecycle;
using Xunit;

namespace NotesShowcase.Tests.ViewModels;

/// <summary>
/// Test-only dialog service whose <see cref="Confirm"/> returns a pre-canned
/// boolean. Used by the Round-4 Important-1 "select then delete clears form"
/// test which needs the in-list delete confirmation to accept.
/// </summary>
file sealed class AlwaysAcceptDialogService : IDialogService
{
    public Task<string?> PickFileToOpen(FileFilter? filter, string? title)
        => Task.FromResult<string?>(null);
    public Task<string?> PickFileToSave(FileFilter? filter, string? title, string? suggestedName)
        => Task.FromResult<string?>(null);
    public Task<bool> Confirm(string message, string? title = null)
        => Task.FromResult(true);
    public Task Notify(string message, string? title = null, NotificationSeverity severity = NotificationSeverity.Info)
        => Task.CompletedTask;
}

public sealed class WorkspaceVMTests
{
    private static WorkspaceVM BuildWorkspace()
    {
        var repo = new InMemoryNoteRepository(
            SeedData.Build(),
            loadAllDelay: TimeSpan.Zero,
            loadNotesDelay: TimeSpan.Zero,
            saveNoteDelay: TimeSpan.Zero,
            addNotebookDelay: TimeSpan.Zero);
        return WorkspaceVM.Builder()
            .Repository(repo)
            .Build();
    }

    private static WorkspaceVM BuildWorkspace(
        IReadOnlyList<NotebookModel> notebooks,
        IReadOnlyList<NoteModel>? notes = null)
    {
        var repo = new InMemoryNoteRepository(
            (notebooks, notes ?? Array.Empty<NoteModel>()),
            loadAllDelay: TimeSpan.Zero,
            loadNotesDelay: TimeSpan.Zero,
            saveNoteDelay: TimeSpan.Zero,
            addNotebookDelay: TimeSpan.Zero);
        return WorkspaceVM.Builder()
            .Repository(repo)
            .Build();
    }

    [Fact]
    public async Task ConstructAsync_loads_notebooks_selects_first_and_populates_notes()
    {
        var ws = BuildWorkspace();
        await ws.ConstructAsync();
        try
        {
            Assert.Equal(ConstructionStatus.Constructed, ws.Status);
            Assert.NotNull(ws.NotebooksRoot.Current);
            Assert.NotEmpty(ws.NotesView.VisibleItems);
            // First root in seed order is nb-work.
            Assert.Equal("nb-work", ws.NotebooksRoot.Current!.Model.Id);
            Assert.All(ws.NotesView.FilteredItems, n => Assert.Equal("nb-work", n.Model.NotebookId));
        }
        finally
        {
            ws.Dispose();
        }
    }

    [Fact]
    public async Task All_six_aggregate_children_are_constructed()
    {
        var ws = BuildWorkspace();
        await ws.ConstructAsync();
        try
        {
            Assert.True(ws.NotebooksRoot.IsConstructed);
            Assert.True(ws.NotesView.IsConstructed);
            Assert.True(ws.NoteForm.IsConstructed);
            Assert.True(ws.StatusBar.IsConstructed);
            Assert.True(ws.Notifications.IsConstructed);
            Assert.True(ws.CapabilityActions.IsConstructed);
        }
        finally
        {
            ws.Dispose();
        }
    }

    [Fact]
    public async Task NewNoteCommand_adds_a_note_in_the_current_notebook()
    {
        var ws = BuildWorkspace();
        await ws.ConstructAsync();
        try
        {
            var nbId = ws.NotebooksRoot.Current!.Model.Id;
            var before = ws.NotesView.FilteredItems.Count;
            ws.NewNoteCommand.Execute(null);
            // NewNote is fire-and-forget; wait for the added note to appear.
            await TestWait.WaitUntilAsync(() => ws.NotesView.FilteredItems.Count > before);
            Assert.True(ws.NotesView.FilteredItems.Count >= before);
            // After rebind, all visible notes belong to the current notebook.
            Assert.All(ws.NotesView.FilteredItems, n => Assert.Equal(nbId, n.Model.NotebookId));
        }
        finally
        {
            ws.Dispose();
        }
    }

    [Fact]
    public async Task SetFocus_triggers_capability_actions_recompute()
    {
        var ws = BuildWorkspace();
        await ws.ConstructAsync();
        try
        {
            // Focus on a notebook: actions list includes "Expand".
            ws.SetFocus(ws.NotebooksRoot.Current!);
            Assert.Contains("Expand", ws.CapabilityActions.Actions.Value.Select(a => a.Label));
            // Focus on a note: actions list switches to note actions.
            var note = ws.NotesView.FilteredItems[0];
            ws.SetFocus(note);
            var labels = ws.CapabilityActions.Actions.Value.Select(a => a.Label).ToList();
            Assert.Contains("Close", labels);
            Assert.DoesNotContain("Expand", labels);
        }
        finally
        {
            ws.Dispose();
        }
    }

    [Fact]
    public async Task Selecting_a_note_updates_capability_focus()
    {
        var ws = BuildWorkspace();
        await ws.ConstructAsync();
        try
        {
            var note = ws.NotesView.FilteredItems[0];
            ws.NotesView.Current = note;

            var labels = ws.CapabilityActions.Actions.Value.Select(a => a.Label).ToList();
            Assert.Contains("Close", labels);
            Assert.Contains("Save", labels);
            Assert.DoesNotContain("Expand", labels);

            ws.NotesView.Current = null;
            var fallback = ws.CapabilityActions.Actions.Value.Select(a => a.Label).ToList();
            Assert.Contains("Expand", fallback);
            Assert.DoesNotContain("Save", fallback);
        }
        finally
        {
            ws.Dispose();
        }
    }

    [Fact]
    public async Task NewNotebookCommand_appends_a_notebook()
    {
        var ws = BuildWorkspace();
        await ws.ConstructAsync();
        try
        {
            var before = ws.NotebooksRoot.All.Count;
            ws.NewNotebookCommand.Execute(null);
            await TestWait.WaitUntilAsync(() => ws.NotebooksRoot.All.Count > before);
            Assert.True(ws.NotebooksRoot.All.Count > before);
        }
        finally
        {
            ws.Dispose();
        }
    }

    [Fact]
    public async Task Selecting_a_note_then_deleting_it_clears_the_form()
    {
        // Round-4 Important-1: when NotesView.Current transitions to null
        // (because the user deletes the selected note), the WorkspaceVM
        // subscription must call NoteForm.Unbind so the right pane does
        // not display ghost data from the just-removed note.
        var repo = new InMemoryNoteRepository(
            SeedData.Build(),
            loadAllDelay: TimeSpan.Zero,
            loadNotesDelay: TimeSpan.Zero,
            saveNoteDelay: TimeSpan.Zero,
            deleteNoteDelay: TimeSpan.Zero,
            addNotebookDelay: TimeSpan.Zero);
        var ws = WorkspaceVM.Builder()
            .Repository(repo)
            // ConfirmationDecorator must accept so the delete actually runs.
            .DialogService(new AlwaysAcceptDialogService())
            .Build();
        await ws.ConstructAsync();
        try
        {
            // Pick a note from the first notebook (seeded nb-work ships
            // with two notes — ConstructAsync already binds to it).
            var note = ws.NotesView.FilteredItems[0];
            ws.NotesView.Current = note;
            // Subscription is synchronous on the immediate-scheduler default
            // — the form should now be bound to the selected note.
            Assert.True(ws.NoteForm.HasBoundNote);
            Assert.Equal(note.Title, ws.NoteForm.Title);

            // Invoke the in-list delete pathway. With the always-accept
            // dialog the ConfirmationDecorator forwards to the inner Task,
            // and NotesViewVM.DeleteNoteAsyncInternal clears Current.
            note.DeleteCommand.Execute(null);
            await TestWait.WaitUntilAsync(() => ws.NotesView.Current is null);

            Assert.Null(ws.NotesView.Current);
            // The form must have been unbound — no ghost data left over.
            Assert.False(ws.NoteForm.HasBoundNote);
            Assert.Equal(string.Empty, ws.NoteForm.Title);
            Assert.Equal(string.Empty, ws.NoteForm.Body);
        }
        finally
        {
            ws.Dispose();
        }
    }

    // ── Pass-6 real-wiring regressions ───────────────────────────────────

    [Fact]
    public async Task Selecting_another_notebook_rebinds_the_notes_view()
    {
        // The tree's TwoWay SelectedItem binding only sets
        // NotebooksRoot.Current — the workspace subscription must do the rest.
        var ws = BuildWorkspace();
        await ws.ConstructAsync();
        try
        {
            var firstId = ws.NotesView.BoundNotebookId;
            var other = ws.NotebooksRoot.Roots.First(nb => nb.Model.Id != firstId);

            ws.NotebooksRoot.Current = other;
            await TestWait.WaitUntilAsync(() =>
            {
                if (ws.NotesView.BoundNotebookId != other.Model.Id)
                {
                    return false;
                }

                try
                {
                    var items = ws.NotesView.FilteredItems.ToArray();
                    return items.Length > 0
                           && items.All(n => n.Model.NotebookId == other.Model.Id);
                }
                catch (InvalidOperationException)
                {
                    return false;
                }
            });

            Assert.Equal(other.Model.Id, ws.NotesView.BoundNotebookId);
            var snapshot = ws.NotesView.FilteredItems.ToArray();
            Assert.All(snapshot, n => Assert.Equal(other.Model.Id, n.Model.NotebookId));
        }
        finally
        {
            ws.Dispose();
        }
    }

    [Fact]
    public async Task Toolbar_commands_fire_CanExecuteChanged_after_construction()
    {
        // Avalonia caches CanExecute() from before construction finished;
        // without a trigger the "+ Note" button stayed permanently disabled.
        var ws = BuildWorkspace();
        var changes = 0;
        ws.NewNoteCommand.CanExecuteChanged += (_, _) => changes++;
        Assert.False(ws.NewNoteCommand.CanExecute(null));

        await ws.ConstructAsync();

        try
        {
            Assert.True(changes > 0,
                "CanExecuteChanged must fire so bound buttons re-evaluate");
            Assert.True(ws.NewNoteCommand.CanExecute(null));
        }
        finally
        {
            ws.Dispose();
        }
    }

    [Fact]
    public void NotebookModel_default_is_not_readonly()
    {
        var nb = new NotebookModel("nb", "Notebook", ParentId: null);

        Assert.False(nb.IsReadOnly);
    }

    [Fact]
    public async Task Capability_add_note_is_disabled_for_readonly_notebook()
    {
        var ws = BuildWorkspace(new[]
        {
            new NotebookModel("nb-readonly", "Archive", ParentId: null, IsReadOnly: true),
        });
        await ws.ConstructAsync();
        try
        {
            Assert.True(ws.NotesView.CurrentNotebookIsReadOnly);
            Assert.False(ws.CapabilityActions.AddNoteCommand.CanExecute(null));
        }
        finally
        {
            ws.Dispose();
        }
    }

    [Fact]
    public async Task Capability_add_note_is_enabled_for_writable_notebook()
    {
        var ws = BuildWorkspace(new[]
        {
            new NotebookModel("nb-rw", "Drafts", ParentId: null),
        });
        await ws.ConstructAsync();
        try
        {
            Assert.False(ws.NotesView.CurrentNotebookIsReadOnly);
            Assert.True(ws.CapabilityActions.AddNoteCommand.CanExecute(null));
        }
        finally
        {
            ws.Dispose();
        }
    }

    [Fact]
    public async Task Deny_reverts_and_republishes_the_draft_surface()
    {
        // The inner FormVM's Deny publishes with sender = FormVM, which the
        // XAML bindings (keyed on NoteFormVM) never observe — the stable
        // DenyCommand must re-emit this VM's own channels.
        var ws = BuildWorkspace();
        await ws.ConstructAsync();
        try
        {
            var note = ws.NotesView.Inner[0];
            ws.NotesView.Current = note;
            var original = ws.NoteForm.Title;
            ws.NoteForm.Title = original + " (edited)";
            Assert.True(ws.NoteForm.IsDirty);

            var raised = new List<string>();
            ws.NoteForm.PropertyChanged += (_, e) => raised.Add(e.PropertyName ?? "");
            ws.NoteForm.DenyCommand.Execute(null);

            Assert.False(ws.NoteForm.IsDirty);
            Assert.Equal(original, ws.NoteForm.Title);
            Assert.Contains(nameof(NoteFormVM.Title), raised);
        }
        finally
        {
            ws.Dispose();
        }
    }

    [Fact]
    public async Task Save_refreshes_the_list_rows_title_and_star()
    {
        var ws = BuildWorkspace();
        await ws.ConstructAsync();
        try
        {
            var note = ws.NotesView.Inner[0];
            ws.NotesView.Current = note;
            ws.NoteForm.Title = "Retitled by test";
            await ws.NoteForm.ApproveAsync();
            await TestWait.WaitUntilAsync(
                () => ws.NotesView.FilteredItems.Any(n => n.Title == "Retitled by test"));

            Assert.Equal("Retitled by test", note.Title);
            Assert.Contains(ws.NotesView.FilteredItems,
                n => n.Title == "Retitled by test");
        }
        finally
        {
            ws.Dispose();
        }
    }

    private sealed class SaveSpyRepository : INoteRepository
    {
        private readonly INoteRepository _inner;
        public int SaveCalls;
        public SaveSpyRepository(INoteRepository inner) => _inner = inner;
        public Task<(IReadOnlyList<NotebookModel> Notebooks, IReadOnlyList<NoteModel> Notes)> LoadAllAsync(CancellationToken ct = default)
            => _inner.LoadAllAsync(ct);
        public Task<IReadOnlyList<NoteModel>> LoadNotesAsync(string notebookId, CancellationToken ct = default)
            => _inner.LoadNotesAsync(notebookId, ct);
        public Task<NoteSearchPage> SearchNotesAsync(string term, string? token, int pageSize, CancellationToken ct = default)
            => _inner.SearchNotesAsync(term, token, pageSize, ct);
        public Task SaveNoteAsync(NoteModel note, CancellationToken ct = default)
        {
            Interlocked.Increment(ref SaveCalls);
            return _inner.SaveNoteAsync(note, ct);
        }
        public Task DeleteNoteAsync(string id, CancellationToken ct = default)
            => _inner.DeleteNoteAsync(id, ct);
        public Task AddNotebookAsync(NotebookModel notebook, CancellationToken ct = default)
            => _inner.AddNotebookAsync(notebook, ct);
        public Task ExportAsync(IReadOnlyList<NotebookModel> notebooks, IReadOnlyList<NoteModel> notes, string path, CancellationToken ct = default)
            => _inner.ExportAsync(notebooks, notes, path, ct);
    }

    [Fact]
    public async Task Capability_save_and_close_act_on_the_focused_note()
    {
        // Pins the OnSave/OnClose wiring in NotesViewVM.ReplaceItems — both
        // capability-bar actions were silent no-ops before pass 6.
        var spy = new SaveSpyRepository(new InMemoryNoteRepository(
            SeedData.Build(),
            loadAllDelay: TimeSpan.Zero,
            loadNotesDelay: TimeSpan.Zero,
            saveNoteDelay: TimeSpan.Zero,
            addNotebookDelay: TimeSpan.Zero));
        var ws = WorkspaceVM.Builder().Repository(spy).Build();
        await ws.ConstructAsync();
        try
        {
            var note = ws.NotesView.Inner[0];
            ws.NotesView.Current = note;

            note.SaveCommand.Execute(null);
            await TestWait.WaitUntilAsync(() => spy.SaveCalls > 0);
            Assert.True(spy.SaveCalls > 0, "capability Save must reach the repository");

            note.CloseCommand.Execute(null);
            Assert.Null(ws.NotesView.Current);
        }
        finally
        {
            ws.Dispose();
        }
    }

    private sealed class FailingLoadRepository : INoteRepository
    {
        private readonly INoteRepository _inner;
        private readonly string _failId;
        public int FailedLoads;
        public FailingLoadRepository(INoteRepository inner, string failId)
        { _inner = inner; _failId = failId; }
        public Task<(IReadOnlyList<NotebookModel> Notebooks, IReadOnlyList<NoteModel> Notes)> LoadAllAsync(CancellationToken ct = default)
            => _inner.LoadAllAsync(ct);
        public Task<IReadOnlyList<NoteModel>> LoadNotesAsync(string notebookId, CancellationToken ct = default)
        {
            if (notebookId == _failId && FailedLoads++ == 0)
            {
                throw new InvalidOperationException("transient repo failure");
            }
            return _inner.LoadNotesAsync(notebookId, ct);
        }
        public Task<NoteSearchPage> SearchNotesAsync(string term, string? token, int pageSize, CancellationToken ct = default)
            => _inner.SearchNotesAsync(term, token, pageSize, ct);
        public Task SaveNoteAsync(NoteModel note, CancellationToken ct = default) => _inner.SaveNoteAsync(note, ct);
        public Task DeleteNoteAsync(string id, CancellationToken ct = default) => _inner.DeleteNoteAsync(id, ct);
        public Task AddNotebookAsync(NotebookModel notebook, CancellationToken ct = default) => _inner.AddNotebookAsync(notebook, ct);
        public Task ExportAsync(IReadOnlyList<NotebookModel> notebooks, IReadOnlyList<NoteModel> notes, string path, CancellationToken ct = default)
            => _inner.ExportAsync(notebooks, notes, path, ct);
    }

    [Fact]
    public async Task A_failed_notebook_bind_does_not_pin_the_selection()
    {
        // Pass-7: a throwing bind must clear _requestedNotebookId so the
        // notebook stays selectable, and the fault must be observed.
        var repo = new FailingLoadRepository(
            new InMemoryNoteRepository(
                SeedData.Build(),
                loadAllDelay: TimeSpan.Zero,
                loadNotesDelay: TimeSpan.Zero,
                saveNoteDelay: TimeSpan.Zero,
                addNotebookDelay: TimeSpan.Zero),
            failId: "nb-personal");
        var ws = WorkspaceVM.Builder().Repository(repo).Build();
        await ws.ConstructAsync();
        try
        {
            var personal = ws.NotebooksRoot.Roots.First(nb => nb.Model.Id == "nb-personal");
            ws.NotebooksRoot.Current = personal;
            await TestWait.WaitUntilAsync(() => repo.FailedLoads >= 1);
            Assert.Equal(1, repo.FailedLoads);
            Assert.NotEqual("nb-personal", ws.NotesView.BoundNotebookId);

            // Re-selecting after a failure must retry (the requested id was
            // cleared). Bounce through another notebook so Current changes.
            ws.NotebooksRoot.Current = ws.NotebooksRoot.Roots.First(nb => nb.Model.Id == "nb-work");
            await TestWait.WaitUntilAsync(() => ws.NotesView.BoundNotebookId == "nb-work");
            ws.NotebooksRoot.Current = personal;
            await TestWait.WaitUntilAsync(() => ws.NotesView.BoundNotebookId == "nb-personal");
            Assert.Equal("nb-personal", ws.NotesView.BoundNotebookId);
        }
        finally
        {
            ws.Dispose();
        }
    }

    [Fact]
    public async Task Construct_with_no_notebooks_still_enables_the_toolbar()
    {
        var repo = new InMemoryNoteRepository(
            (Array.Empty<NotebookModel>(), Array.Empty<NoteModel>()),
            loadAllDelay: TimeSpan.Zero,
            loadNotesDelay: TimeSpan.Zero,
            saveNoteDelay: TimeSpan.Zero,
            addNotebookDelay: TimeSpan.Zero);
        var ws = WorkspaceVM.Builder().Repository(repo).Build();
        await ws.ConstructAsync();
        try
        {
            Assert.True(ws.NewNotebookCommand.CanExecute(null));
            Assert.False(ws.NewNoteCommand.CanExecute(null), "no current notebook yet");
        }
        finally
        {
            ws.Dispose();
        }
    }

    [Fact]
    public async Task Export_command_writes_the_workspace_through_the_dialog_path()
    {
        var path = Path.Combine(Path.GetTempPath(), $"vmx-export-{Guid.NewGuid():N}.json");
        var ws = WorkspaceVM.Builder()
            .Repository(new InMemoryNoteRepository(
                SeedData.Build(),
                loadAllDelay: TimeSpan.Zero,
                loadNotesDelay: TimeSpan.Zero,
                saveNoteDelay: TimeSpan.Zero,
                addNotebookDelay: TimeSpan.Zero,
                exportDelay: TimeSpan.Zero))
            .DialogService(new SaveDialogService(path))
            .Build();
        await ws.ConstructAsync();
        try
        {
            ws.ExportCommand.Execute(null);
            await TestWait.WaitUntilAsync(() => File.Exists(path));
            Assert.True(File.Exists(path), "export must write through the picked path");
        }
        finally
        {
            ws.Dispose();
            if (File.Exists(path)) File.Delete(path);
        }
    }

    private sealed class SaveDialogService : IDialogService
    {
        private readonly string _path;
        public SaveDialogService(string path) => _path = path;
        public Task<string?> PickFileToOpen(FileFilter? filter, string? title)
            => Task.FromResult<string?>(null);
        public Task<string?> PickFileToSave(FileFilter? filter, string? title, string? suggestedName)
            => Task.FromResult<string?>(_path);
        public Task<bool> Confirm(string message, string? title = null)
            => Task.FromResult(true);
        public Task Notify(string message, string? title = null, NotificationSeverity severity = NotificationSeverity.Info)
            => Task.CompletedTask;
    }

    [Fact]
    public async Task Destruct_cascades_to_all_six_children()
    {
        var ws = BuildWorkspace();
        await ws.ConstructAsync();
        ws.Destruct();
        Assert.Equal(ConstructionStatus.Destructed, ws.NotebooksRoot.Status);
        Assert.Equal(ConstructionStatus.Destructed, ws.NotesView.Status);
        Assert.Equal(ConstructionStatus.Destructed, ws.NoteForm.Status);
        Assert.Equal(ConstructionStatus.Destructed, ws.StatusBar.Status);
        Assert.Equal(ConstructionStatus.Destructed, ws.Notifications.Status);
        Assert.Equal(ConstructionStatus.Destructed, ws.CapabilityActions.Status);
        ws.Dispose();
    }
}

File: examples/swift/notes-showcase/Tests/NotesShowcaseTests/WorkspaceVMTests.swift
//
// WorkspaceVMTests — scenario tests for WorkspaceVM (composition root).
//
// Ports examples/csharp/avalonia/NotesShowcase.Tests/ViewModels/WorkspaceVMTests.cs.
// No conformance-ID markers (scenario IDs live in THEME-00x only).
//
// Determinism strategy:
//   • Zero repo delays → async repo calls return immediately.
//   • ImmediateDispatcher.INSTANCE → scheduleForeground runs inline.
//   • `waitUntil` drains fire-and-forget Tasks (addNotebook, newNote, etc.)
//     by yielding until the condition holds or a retry cap is hit.
//
import XCTest
import Combine
import VMx
@testable import NotesShowcaseCore

// MARK: - Test dialog services

/// Dialog that silently accepts every confirmation and returns `nil` for files.
private final class AlwaysAcceptDialogService: DialogService {
    func pickFileToOpen(filter: FileFilter?, title: String?) async -> String? { nil }
    func pickFileToSave(filter: FileFilter?, title: String?, suggestedName: String?) async -> String? { nil }
    func confirm(_ message: String, title: String?) async -> Bool { true }
    func notify(_ message: String, title: String?, severity: NotificationSeverity) async {}
}

/// Dialog that always returns a pre-configured path from `pickFileToSave`.
private final class SaveDialogService: DialogService {
    private let _path: String
    init(path: String) { _path = path }
    func pickFileToOpen(filter: FileFilter?, title: String?) async -> String? { nil }
    func pickFileToSave(filter: FileFilter?, title: String?, suggestedName: String?) async -> String? { _path }
    func confirm(_ message: String, title: String?) async -> Bool { true }
    func notify(_ message: String, title: String?, severity: NotificationSeverity) async {}
}

// MARK: - WorkspaceVMTests

final class WorkspaceVMTests: XCTestCase {

    // MARK: - Helpers

    /// Builds a zero-delay workspace wired to the standard seed.
    private func buildWorkspace(
        dialogService: (any DialogService)? = nil
    ) throws -> WorkspaceVM {
        let repo = InMemoryNoteRepository(
            seed: SeedData.build(),
            loadAllDelay: 0,
            loadNotesDelay: 0,
            saveNoteDelay: 0,
            deleteNoteDelay: 0,
            addNotebookDelay: 0,
            exportDelay: 0
        )
        var b = WorkspaceVM.builder().repository(repo)
        if let ds = dialogService { b = b.dialogService(ds) }
        return try b.build()
    }

    /// Polls until `condition` returns `true` or the retry cap is reached.
    ///
    /// Uses `Task.sleep(nanoseconds: 1_000_000)` (1 ms) rather than
    /// `Task.yield()` between checks so the cooperative-pool threads have
    /// real wall-clock time to drain nested async chains (outer spawn →
    /// bindNotesObserved → bindTo inner Task → actor hop).  Pure
    /// `Task.yield()` from the main actor only reschedules other
    /// *main-actor* work; cooperative-pool tasks may not advance within
    /// the allowed attempt window on loaded CI runners.
    private func waitUntil(_ condition: @escaping () -> Bool, attempts: Int = 100) async {
        for _ in 0..<attempts {
            if condition() { return }
            try? await Task.sleep(nanoseconds: 1_000_000)   // 1 ms per iteration
        }
    }

    // MARK: - Construction

    func testConstructAsync_loadsNotebooks_selectsFirst_populatesNotes() async throws {
        let ws = try buildWorkspace()
        try await ws.constructAsync()
        defer { ws.dispose() }

        XCTAssertEqual(.constructed, ws.status)
        XCTAssertNotNil(ws.notebooksRoot.current)
        XCTAssertFalse(ws.notesView.visibleItems.isEmpty)
        // First root in seed order is nb-work.
        XCTAssertEqual("nb-work", ws.notebooksRoot.current?.model.id)
        XCTAssertTrue(
            ws.notesView.filteredItems.allSatisfy { $0.model.notebookId == "nb-work" }
        )
    }

    func testAllSixAggregateChildren_areConstructed() async throws {
        let ws = try buildWorkspace()
        try await ws.constructAsync()
        defer { ws.dispose() }

        XCTAssertTrue(ws.notebooksRoot.isConstructed)
        XCTAssertTrue(ws.notesView.isConstructed)
        XCTAssertTrue(ws.noteForm.isConstructed)
        XCTAssertTrue(ws.statusBar.isConstructed)
        XCTAssertTrue(ws.notifications.isConstructed)
        XCTAssertTrue(ws.capabilityActions.isConstructed)
    }

    func testTheme_isConstructedAlongsideAggregate() async throws {
        let ws = try buildWorkspace()
        try await ws.constructAsync()
        defer { ws.dispose() }

        XCTAssertTrue(ws.theme.isConstructed)
    }

    // MARK: - Notebook selection rebinds notes view

    func testSelectingAnotherNotebook_rebindsNotesView() async throws {
        let ws = try buildWorkspace()
        try await ws.constructAsync()
        defer { ws.dispose() }

        let firstId = ws.notesView.boundNotebookId
        // Pick a root different from the currently bound one.
        guard let other = ws.notebooksRoot.roots.first(where: { $0.model.id != firstId }) else {
            XCTFail("Expected at least two root notebooks in seed data")
            return
        }

        ws.notebooksRoot.current = other
        await waitUntil { ws.notesView.boundNotebookId == other.model.id }

        XCTAssertEqual(other.model.id, ws.notesView.boundNotebookId)
        XCTAssertTrue(
            ws.notesView.filteredItems.allSatisfy { $0.model.notebookId == other.model.id }
        )
    }

    // MARK: - Note form wiring

    func testSelectingANote_bindsTheForm() async throws {
        let ws = try buildWorkspace()
        try await ws.constructAsync()
        defer { ws.dispose() }

        let note = ws.notesView.filteredItems.first
        XCTAssertNotNil(note, "Expected notes after constructAsync")
        ws.notesView.current = note

        XCTAssertTrue(ws.noteForm.hasBoundNote)
        XCTAssertEqual(note?.model.title, ws.noteForm.title)
    }

    func testClearingCurrentNote_unbindsTheForm() async throws {
        // Round-4 Important-1: when Current becomes nil (e.g. after delete),
        // the form must be unbound so no ghost data lingers in the right pane.
        let ws = try buildWorkspace(dialogService: AlwaysAcceptDialogService())
        try await ws.constructAsync()
        defer { ws.dispose() }

        let note = ws.notesView.filteredItems.first
        XCTAssertNotNil(note)
        ws.notesView.current = note

        XCTAssertTrue(ws.noteForm.hasBoundNote)

        // Delete the selected note (AlwaysAccept confirms the prompt).
        note?.deleteCommand.execute()
        await waitUntil { ws.notesView.current == nil }

        XCTAssertNil(ws.notesView.current)
        XCTAssertFalse(ws.noteForm.hasBoundNote)
        XCTAssertEqual("", ws.noteForm.title)
        XCTAssertEqual("", ws.noteForm.body)
    }

    // MARK: - onSaved → refreshNote wiring

    func testSave_refreshesListRowTitleAndStar() async throws {
        let ws = try buildWorkspace()
        try await ws.constructAsync()
        defer { ws.dispose() }

        let note = ws.notesView.inner.at(0)
        ws.notesView.current = note
        ws.noteForm.title = "Retitled by test"
        try await ws.noteForm.approveAsync()

        await waitUntil {
            ws.notesView.filteredItems.contains { $0.title == "Retitled by test" }
        }

        XCTAssertEqual("Retitled by test", note.title)
        XCTAssertTrue(
            ws.notesView.filteredItems.contains { $0.title == "Retitled by test" }
        )
    }

    // MARK: - Commands

    func testNewNotebookCommand_appendsANotebook() async throws {
        let ws = try buildWorkspace()
        try await ws.constructAsync()
        defer { ws.dispose() }

        let before = ws.notebooksRoot.all.count
        ws.newNotebookCommand.execute()
        await waitUntil { ws.notebooksRoot.all.count > before }

        XCTAssertGreaterThan(ws.notebooksRoot.all.count, before)
    }

    func testNewNoteCommand_addsNoteInCurrentNotebook() async throws {
        let ws = try buildWorkspace()
        try await ws.constructAsync()
        defer { ws.dispose() }

        let nbId = ws.notebooksRoot.current?.model.id
        XCTAssertNotNil(nbId)
        let before = ws.notesView.filteredItems.count

        ws.newNoteCommand.execute()
        await waitUntil { ws.notesView.filteredItems.count > before }

        XCTAssertGreaterThan(ws.notesView.filteredItems.count, before)
        if let nbId = nbId {
            XCTAssertTrue(
                ws.notesView.filteredItems.allSatisfy { $0.model.notebookId == nbId }
            )
        }
    }

    func testToolbarCommands_fireCanExecuteChanged_afterConstruction() async throws {
        // Avalonia (and SwiftUI) caches CanExecute from before construction;
        // without a command-trigger fire the "+ Note" button stays disabled.
        let ws = try buildWorkspace()
        var changes = 0
        let sub = ws.newNoteCommand.canExecuteChanged.sink { changes += 1 }

        XCTAssertFalse(ws.newNoteCommand.canExecute())  // Before construction: no notebook yet.

        try await ws.constructAsync()
        defer { ws.dispose(); sub.cancel() }

        XCTAssertGreaterThan(changes, 0,
            "canExecuteChanged must fire so bound UI buttons re-evaluate")
        XCTAssertTrue(ws.newNoteCommand.canExecute())
    }

    func testConstruct_withNoNotebooks_stillEnablesNewNotebookButton() async throws {
        let repo = InMemoryNoteRepository(
            seed: (notebooks: [], notes: []),
            loadAllDelay: 0,
            loadNotesDelay: 0,
            saveNoteDelay: 0
        )
        let ws = try WorkspaceVM.builder().repository(repo).build()
        try await ws.constructAsync()
        defer { ws.dispose() }

        XCTAssertTrue(ws.newNotebookCommand.canExecute())
        XCTAssertFalse(ws.newNoteCommand.canExecute(),
            "newNote requires a current notebook — none selected yet")
    }

    // MARK: - setFocus / capability actions

    func testSetFocus_triggersCapabilityActionsRecompute() async throws {
        let ws = try buildWorkspace()
        try await ws.constructAsync()
        defer { ws.dispose() }

        // Focus on a notebook: actions list includes "Expand" (Expandable).
        let nb = ws.notebooksRoot.current
        XCTAssertNotNil(nb)
        ws.setFocus(nb!)
        XCTAssertTrue(
            try ws.capabilityActions.actions.value.contains { $0.label == "Expand" }
        )

        // Focus on a note: actions switch to note-specific actions.
        let noteVM = ws.notesView.filteredItems.first
        XCTAssertNotNil(noteVM)
        ws.setFocus(noteVM!)
        let labels = try ws.capabilityActions.actions.value.map(\.label)
        XCTAssertTrue(labels.contains("Close"))
        XCTAssertFalse(labels.contains("Expand"))
    }

    // MARK: - Deny / revert

    func testDenyCommand_revertsAndRepublishesDraftSurface() async throws {
        let ws = try buildWorkspace()
        try await ws.constructAsync()
        defer { ws.dispose() }

        let note = ws.notesView.inner.at(0)
        ws.notesView.current = note
        let original = ws.noteForm.title
        ws.noteForm.title = original + " (edited)"
        XCTAssertTrue(ws.noteForm.isDirty)

        ws.noteForm.denyCommand.execute()

        XCTAssertFalse(ws.noteForm.isDirty)
        XCTAssertEqual(original, ws.noteForm.title)
    }

    // MARK: - Lifecycle cascade

    func testDestruct_cascadesToAllSixChildren() async throws {
        let ws = try buildWorkspace()
        try await ws.constructAsync()
        try ws.destruct()
        defer { ws.dispose() }

        XCTAssertEqual(.destructed, ws.notebooksRoot.status)
        XCTAssertEqual(.destructed, ws.notesView.status)
        XCTAssertEqual(.destructed, ws.noteForm.status)
        XCTAssertEqual(.destructed, ws.statusBar.status)
        XCTAssertEqual(.destructed, ws.notifications.status)
        XCTAssertEqual(.destructed, ws.capabilityActions.status)
    }

    // MARK: - Export

    func testExportCommand_writesWorkspaceThroughDialogPath() async throws {
        let path = (NSTemporaryDirectory() as NSString)
            .appendingPathComponent("vmx-export-\(UUID().uuidString).json")
        let ws = try WorkspaceVM.builder()
            .repository(InMemoryNoteRepository(
                seed: SeedData.build(),
                loadAllDelay: 0,
                loadNotesDelay: 0,
                saveNoteDelay: 0,
                exportDelay: 0
            ))
            .dialogService(SaveDialogService(path: path))
            .build()
        try await ws.constructAsync()
        defer {
            ws.dispose()
            try? FileManager.default.removeItem(atPath: path)
        }

        ws.exportCommand.execute()
        await waitUntil { FileManager.default.fileExists(atPath: path) }

        XCTAssertTrue(
            FileManager.default.fileExists(atPath: path),
            "export must write through the picked path"
        )
    }

    // MARK: - Builder

    func testBuilder_throwsWhenRepositoryIsMissing() {
        XCTAssertThrowsError(try WorkspaceVM.builder().build())
    }

    func testBuilder_usesDefaultsForOptionalServices() throws {
        let repo = InMemoryNoteRepository(
            seed: (notebooks: [], notes: []),
            loadAllDelay: 0,
            loadNotesDelay: 0,
            saveNoteDelay: 0
        )
        let ws = try WorkspaceVM.builder().repository(repo).build()
        // No crash when built with only the required field.
        XCTAssertNotNil(ws)
        ws.dispose()
    }
}
